Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp

```diff
@@ -0,0 +1,386 @@
+#include "MediaUtilities.h"
+#include "PlatformRawAudioDataCocoa.h"
+#include "WebMAudioUtilitiesCocoa.h"
+
+#include <CoreAudio/CoreAudioTypes.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/TZoneMallocInlines.h>
+#include <wtf/ThreadSafeRefCounted.h>
+#include <wtf/WorkQueue.h>
+#include <wtf/text/MakeString.h>
+
+#include <pal/cf/CoreMediaSoftLink.h>
+
+namespace WebCore {
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(AudioEncoderCocoa);
+
+class InternalAudioEncoderCocoa : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<InternalAudioEncoderCocoa, WTF::DestructionThread::Main> {
+    WTF_MAKE_TZONE_ALLOCATED_INLINE(InternalAudioEncoderCocoa);
+    WTF_MAKE_NONCOPYABLE(InternalAudioEncoderCocoa);
+
+public:
+    using Config = AudioEncoder::Config;
+    using EncodePromise = AudioEncoder::EncodePromise;
+
+    static Ref<InternalAudioEncoderCocoa> create(const Config& config, AudioEncoder::DescriptionCallback&& descriptionCallback, AudioEncoder::OutputCallback&& outputCallback)
+    {
+        return adoptRef(*new InternalAudioEncoderCocoa(config, WTFMove(descriptionCallback), WTFMove(outputCallback)));
+    }
+    ~InternalAudioEncoderCocoa() = default;
+
+    String initialize(const String& codecName);
+    Ref<AudioEncoder::EncodePromise> encode(AudioEncoder::RawFrame&&);
+    Ref<GenericPromise> flush();
+    void reset();
+    void close();
+
+    static WorkQueue& queueSingleton() { return AudioDecoderCocoa::queueSingleton(); }
+
+private:
+    InternalAudioEncoderCocoa(const AudioEncoder::Config&, AudioEncoder::DescriptionCallback&&, AudioEncoder::OutputCallback&&);
+    static void compressedAudioOutputBufferCallback(void*, CMBufferQueueTriggerToken);
+    Ref<AudioSampleBufferConverter> converter() const { return *m_converter; }
+    void processEncodedOutputs();
+    AudioEncoder::ActiveConfiguration activeConfiguration(CMSampleBufferRef) const;
+    Vector<uint8_t> generateDecoderDescriptionFromSample(CMSampleBufferRef) const;
+
+    AudioEncoder::DescriptionCallback m_descriptionCallback;
+    AudioEncoder::OutputCallback m_outputCallback;
+    RefPtr<AudioSampleBufferConverter> m_converter;
+
+    bool m_isClosed WTF_GUARDED_BY_CAPABILITY(queueSingleton()) { false };
+    const AudioEncoder::Config m_config;
+    String m_codecString; // set during initialisation, const after.
+    FourCharCode m_codec; // set during initialisation, const after.
+    std::optional<AudioStreamDescription::PCMFormat> m_pcmFormat; // set during initialisation, const after.
+    std::optional<CAAudioStreamDescription> m_outputDescription; // set during initialisation, const after.
+    std::optional<MediaTime> m_frameDuration; // set during initialisation, const after.
+    std::optional<unsigned> m_complexity; // set during initialisation, const after.
+    std::optional<unsigned> m_packetlossperc; // set during initialisation, const after.
+    std::optional<bool> m_useinbandfec; // set during initialisation, const after.
```

Comment:
It would be better to have a function that returns all these parameters and then we create the InternalAudioEncoderCocoa.
That way, we can make them const.

Reply:
that was more effort than worth imho.
but done